addOns/graphql/src/main/java/org/zaproxy/addon/graphql/GraphQlGenerator.java

```diff
@@ -0,0 +1,329 @@
+import graphql.schema.GraphQLList;
+import graphql.schema.GraphQLNamedOutputType;
+import graphql.schema.GraphQLNamedType;
+import graphql.schema.GraphQLNonNull;
+import graphql.schema.GraphQLObjectType;
+import graphql.schema.GraphQLScalarType;
+import graphql.schema.GraphQLSchema;
+import graphql.schema.GraphQLType;
+import graphql.schema.GraphQLTypeUtil;
+import graphql.schema.GraphQLUnionType;
+import graphql.schema.idl.SchemaParser;
+import graphql.schema.idl.UnExecutableSchemaGenerator;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import org.apache.log4j.Logger;
+import org.parosproxy.paros.control.Control;
+import org.zaproxy.zap.extension.spider.ExtensionSpider;
+import org.zaproxy.zap.model.ValueGenerator;
+
+public class GraphQlGenerator {
+
+    private static final Logger LOG = Logger.getLogger(GraphQlGenerator.class);
+    private static final int MAX_DEPTH = 5;
+    private final Requestor requestor;
+    private GraphQLSchema schema;
+
+    protected enum RequestType {
+        QUERY,
+        MUTATION,
+        SUBSCRIPTION
+    };
+
+    public GraphQlGenerator(String sdl) {
+        this(sdl, null);
+    }
+
+    public GraphQlGenerator(String sdl, Requestor requestor) {
+        schema = UnExecutableSchemaGenerator.makeUnExecutableSchema(new SchemaParser().parse(sdl));
+        this.requestor = requestor;
+    }
+
+    public void checkServiceMethods() {
+        String query = "{__schema{types{name kind description}}}";
+        requestor.sendQueryByGet(query);
+        requestor.sendQueryByGraphQlPost(query);
+        requestor.sendQuery(query);
+    }
+
+    public String generateFull(RequestType requestType) {
+        StringBuilder query = new StringBuilder();
+        GraphQLObjectType object = null;
+        switch (requestType) {
+            case QUERY:
+                object = schema.getQueryType();
+                break;
+            case MUTATION:
+                query.append("mutation ");
+                object = schema.getMutationType();
+                break;
+            case SUBSCRIPTION:
+                query.append("subscription ");
+                object = schema.getSubscriptionType();
+                break;
+        }
+        generate(query, object, 0);
+        return query.toString();
+    }
+
+    public void sendFull(RequestType requestType) {
+        switch (requestType) {
+            case QUERY:
+                StringBuilder query = new StringBuilder();
+                generate(query, schema.getQueryType(), 0);
+                requestor.sendQuery(query.toString());
+                break;
+            case MUTATION:
+                StringBuilder mutation = new StringBuilder("mutation ");
+                generate(mutation, schema.getMutationType(), 0);
+                requestor.sendQuery(mutation.toString());
+                break;
+            case SUBSCRIPTION:
+                StringBuilder subscription = new StringBuilder("subscription ");
```

Comment:
This is the second use of these strings, specifically with a trailing space. It would probably be best to extract constants for them.